### ISSUE TITLE
Request monitoring data using jetstream endpoints

### ISSUE THREAD

**Issue by charbonnierg** [mdawar/nats-dashboard#4]
This application relies on the NATS HTTP monitoring endpoint to fetch data. Would you consider adding support for using websocket and [NATS endpoints (`$SYS.REQ.SERVER.PING.*`)](https://docs.nats.io/running-a-nats-service/configuration/sys_accounts#system-account) ? 

As stated in the documentation:

> [Monitoring endpoints](https://github.com/running-a-nats-service/configuration/monitoring) as listed in the table below are accessible as system services using the following subject pattern:
$SYS.REQ.SERVER.<id>.<endpoint-name> (request server monitoring endpoint corresponding to endpoint name.)

And if so, how would you consider adding support for authentication/authorization ?

I love this project, but cannot use it as much as I'd want to in practice because the SPA must run somewhere where NATS HTTP endpoint is reachable, which means local network only most of the time.

**Comment by mdawar** [comment]
Yes you're totally right, right now we can only monitor local instances or the demo NATS server, which is publicly exposed (Not recommended for production servers).

From the [docs](https://docs.nats.io/running-a-nats-service/configuration/monitoring):

> Warning: nats-server does not have authentication/authorization for the monitoring endpoint. When you plan to open your nats-server to the internet make sure to not expose the monitoring port as well. By default monitoring binds to every interface 0.0.0.0 so consider setting monitoring to localhost or have appropriate firewall rules.

Unfortunately I don't have any plans on adding support for WebSockets and monitoring using the system endpoints (I don't have much time to dedicate right now).

But I also would like to monitor my production NATS servers without exposing the monitoring port, and I was thinking about that but I still haven't implemented anything yet, when I do so I'll make sure to add this info to the README.

What I can think of right now is using a reverse proxy with authentication (Caddy, Nginx, ...) and then we can add support for prompting the user for authentication on the first request when we get a 401 status.

For example a Caddy server running alongside a NATS server which we can use to reverse proxy to the monitoring endpoint with automatic TLS cert and authentication.

I think this solution keeps the project simple and does not over complicate things.

What do you think?

**Comment by charbonnierg** [comment]
> Unfortunately I don't have any plans on adding support for WebSockets and monitoring using the system endpoints (I don't have much time to dedicate right now).

I understand, as it would be quite the effort. Also, I'm not sure the current UI would work well,  because a single URL may be used to inquire about many servers. Instead of only providing a URL, a user would need to first provide a URL, let the application discover servers, and then select the server to monitor I guess ?

> What I can think of right now is using a reverse proxy with authentication (Caddy, Nginx, ...) and then we can add support for prompting the user for authentication on the first request when we get a 401 status.
For example a Caddy server running alongside a NATS server which we can use to reverse proxy to the monitoring endpoint with automatic TLS cert and authentication.

That seems to be the best solution when using the HTTP endpoint, but I don't really see what this project can bring which does not already exist somewhere else in the community ? For example, I used https://github.com/oauth2-proxy/oauth2-proxy in the past to protect arbitrary endpoints served by caddy. Do you have something in mind ?

**Comment by mdawar** [comment]
Actually this app is nothing more than a tool similar to [nats-top](https://github.com/nats-io/nats-top), but running on the web, that's it.

Some use cases are better served with the prometheus exporter and grafana (with alerts if needed), others would benefit from [Synadia control plane](https://www.synadia.com/control-plane), and also I've seen on twitter a NATS GUI called [Quaze](https://qaze.app/).

And yeah I think I'm going to use a reverse proxy with auth (like oauth2-proxy) for my own use case and I'll document that in the README.

Anyway if you really want to create something better, you can always fork the project and continue with your plans. I honestly don't have the time to do so, sorry.

**Comment by charbonnierg** [comment]
> Anyway if you really want to create something better, you can always fork the project and continue with your plans. I honestly don't have the time to do so, sorry.

I totally understand, moreover I think that your app is much more than nats-top as you've built a great UI which is beautiful and easy to use.  Also the fact that this is a PWA, and that there is no backend involved is really cool.

> Anyway if you really want to create something better, you can always fork the project and continue with your plans. I honestly don't have the time to do so, sorry.

I've never used Astro before, neither did I use solid, or tanstack query, otherwise I would have tried an implementation before opening this issue. If I want to give it a try, do you recommend continuing to use tanstack query but changing the query functions to use NATS request/reply (I don"t even know if it's possible, but the `queries.ts` file under `lib` makes me think that it is ) ? Also, would you recommend to create a single NATS connection (for example under a context) and use it everywhere ?

Anyway, thanks for your time and your project

**Comment by mdawar** [comment]
Yeah sure keep using Tanstask query, it's very versatile, it doesn't care where you're getting the data from, it doesn't make the fetch requests, you can provide any function that returns the data and it does the caching, the invalidation and much more.
I tried to document all the code, so it should be easy to work with.

I haven't used the NATS websocket client very much, I've only tried it once and it was easy to set up and work with, I'm assuming you're going to use it to create the requests, you can create a single connection at the top level component and share it maybe with a context, and like I previously said, any function that can return the data can be used with Tanstack query.

And about Solid, if you know React then it's quite easy to work with Solid, actually I learned solid while creating this project, and I found it much better to work with than React, also you don't have to worry about re-renders of your components, they're only called once (They're regular functions, nothing special).

I hope this helps.

**Comment by charbonnierg** [comment]
I pushed a new commit on my fork. Basically, I'm very satisfied with what it allows me to do: I can connect to a single server, either without auth, with user/pass, with token or with creds file, and monitor different servers, but the UX is poor:

- The first change was to create a new NATS connection when user toggles the active state.
  - 🚨 In case the connection fails, the active state is toggled back to false, but error is not displayed to user (so user is not aware of authorization errors for example).

- The second change was to introduce additional settings to configure the NATS connection:
  <img width="386" alt="image" src="https://github.com/mdawar/nats-dashboard/assets/81654818/e5a3657d-a130-4d83-b4fa-862cc5b5aa77">

  - ✅ If an HTTP url is used, no additional settings are displayed (because they would not be used anyway)
  - ✅ If an URL starting with `ws://` or `wss://` is used, `JSONP` settings is hidden (because it's not used anyway)
  - ✅ If one auth is used, others are cleared (e.g., if a token was previously used, and a username/password is used, token is cleared and same goes for creds file).
  - 🚨 credentials are "lost" on each page refresh, and I'm reluctant to store credentials because I'm no security expert.

- Finally I added the possibility to select a different server, and because I wanted a quick-and-dirty solution I reused the app settings:
  <img width="382" alt="image" src="https://github.com/mdawar/nats-dashboard/assets/81654818/3ef400e0-6879-40da-864e-0fbfa73ca697">

  - 🚨 I dont think that this should be part of the app settings, but rather part of the main page. I was reluctant to change the main page as this feature would not be available when using HTTP endpoints.

The most challenging parts are:
- What to do with credentials ? Should users be asked to enter credentials on each refresh or is it acceptable to prompt user for a single password and encode/decode credentials using user password somewhere in local storage ?
- What would be the best UX for server selection ? Maybe simply clicking on the server name on the dashboard and displaying a select input ?

Sorry if I'm bothering you, I understand that you don't have time for this project at the moment so I'm not expecting anything but any insight would be much appreciated :)

**Comment by mdawar** [comment]
Well, I can't really answer these questions for sure, because honestly I haven't tried any of the NATS authentication methods yet, I'm using it in a private network without authentication.

And yes storing the user credentials in the browser is always risky, but if you really want to do it, you can store it in [sessionStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage) which will be cleared when the browser tab is closed.

I'm sorry I couldn't provide you with precise answers.
